### PR TITLE
otk/command: implement "validate" command

### LINKED ISSUE #79: Add an `otk validate` subcommand
Let's add an `otk validate` subcommand which validates a given omnifest but produces no output. The exit code should be `0` for a valid omnifest and `1` for a non-valid omnifest.

### PR DESCRIPTION
**Pull request by schuellerf**
I also fixed the behavior of "output" as it was expected to be a path - which is wrong I guess
fixes #79

### CODE REVIEW

**Comment by schuellerf** [comment]
as I ran into the linter in github - I also integrated ruff into pre-commit now.

please comment on those:
 - should this be a separate PR?
 - should we keep the "--fix" in pre-commit for `ruff`?

**Comment by supakeen** [comment]
> as I ran into the linter in github - I also integrated ruff into pre-commit now.
> 
> please comment on those:
> 
>     * should this be a separate PR?
> 
>     * should we keep the "--fix" in pre-commit for `ruff`?

I'm fine with having this in the same PR, let's not overcomplicate since I agree with both additions.

There's just an architectural remark I have. I don't quite like `validate` calling the `compile` subcommand and I'd prefer having them as separate commands, we can move the common code to a helper function.

**Review by supakeen — CHANGES_REQUESTED** [review]
I already wrote this in a normal comment; I'd like to keep the `compile` and `validate` commands separate and move their function bodies to a common helper.

**Comment by mvo5** [comment]
Fwiw, I would be in favor of splitting purely whitespace changes (like the last (two?) commit(s) that just does a bunch of ruff formatting) into a separate PR. I think it's distracting otherwise for the reviewer (and for people looking at the PR history)

File: test/test_command.py [hunk 1]
```diff
@@ -1,65 +1,46 @@
...
+    with pytest.raises(SystemExit) as sys_exit:
+        p.parse_args([])
+    assert sys_exit.type == SystemExit
+    assert sys_exit.value.code == 2
+
+    captured_stderr = capsys.readouterr().err
+    assert "the following arguments are required: command" in captured_stderr
+
+
+@pytest.mark.parametrize("command,results,sys_exit_code,sys_exit_message", [
```
Comment (mvo5) [inline test/test_command.py:17]: I like this!

File: test/test_command.py [hunk 2]
```diff
@@ -1,65 +1,46 @@
...
-    r = p.parse_args(["compile", "-o", "file.yaml"])
-    assert r.command == "compile"
-    assert r.input is None
-    assert r.output == "file.yaml"
-
-    r = p.parse_args(["compile", "-o", "file.yaml", "foo.yaml"])
-    assert r.command == "compile"
-    assert r.input == "foo.yaml"
-    assert r.output == "file.yaml"
+    if sys_exit_code:
```
Comment (mvo5) [inline test/test_command.py:35]: (super nitpick) Given how small the body is, I wonder if it's worthwhile to split the test into two, one "happy" and one "sad" case. This way the inputs are slightly smaller and the pytest.mark.parametriz might be slightly easier to read (or maybe not) - in any case, more of a thought then aynthing else, fine as is.

File: test/test_compile.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
+import argparse
+import os.path
+import pytest
+
+from otk.command import compile
+
+
+@pytest.mark.parametrize(
+    "parser, arguments, dry_run, input_data, output_data, sys_exit_code, log_message",
```
Comment (mvo5) [inline test/test_compile.py:9]: I wonder if this one can be split a bit more, it feels a bit hard to read and the test body has a bunch of "if" conditions inside, could this be split into multiple smaller tests maybe somehow? Maybe a "happy" and "sad" might already help?

File: test/test_compile.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    [
+        # "GENERATE" is a placeholder to append "tmp_path" from pytest
+        # dry_run must not generate an output
+        (
+            None,
```
Comment (mvo5) [inline test/test_compile.py:14]: This seems to be "None" for all inputs? Or did I just not look careful enough?

File: test/test_compile.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  "test": 1,
+  "version": "2",
+  "sources": {}
+}""",
+            0,
+            None,
+        ),
+    ],
+)
+def test_validate(
```
Comment (mvo5) [inline test/test_compile.py:64]: This is called "test_validate" but the test is called test_compile.py and the function under test is "compile()" too, is this intentional?

**Review by mvo5 — COMMENTED** [review]
Thanks for the PR! Lots to like here! I added some questions and ideas inline for your consideration. Note that I only reviewed the code changes (first commit).

File: test/test_compile.py (hunk 4 above)
Comment (schuellerf) [reply test/test_compile.py:14]: Well the parameter itself is not even used in the function - yes, I could drop it from the test for now...

File: src/otk/command.py [hunk 6]
```diff
@@ -116,13 +118,22 @@ def compile(
...
+    if not dry_run:
+        dst.write_text(
+            target_registry.get(kind, CommonTarget)().as_string(spec, tree)
+        )
 
     return 0
 
 
+def compile(parser: argparse.ArgumentParser, arguments: argparse.Namespace) -> int:
+    return _process(parser, arguments, False)
```
Comment (mvo5) [inline src/otk/command.py:129]: Maybe change this to `return _process(parser, arguments, dry_runFalse)` (and similar below)? This way it becomes easier to immediately see what the "true/false" mean in here?

File: test/test_compile.py (hunk 4 above)
Comment (schuellerf) [reply test/test_compile.py:14]: @supakeen I can as well remove the parameter "parser" from `compile` (and `validate` and `_process()` how it is called here)
Probably this will never be used?

File: test/test_compile.py (hunk 5 above)
Comment (schuellerf) [reply test/test_compile.py:64]: Nope - just started with validate tests and forgot to rename when I saw we could as well test both

File: test/test_compile.py (hunk 4 above)
Comment (schuellerf) [reply test/test_compile.py:14]: @mvo5 btw. also `arguments.input` is always `"GENERATE"` because I wanted to test the stdin case - but I don't know how to mock `/dev/stdin`. I could mock `sys.stdin` but not `/dev/stdin` - any ideas? 🤔
(without rewriting the whole functionality of otk 😅)

Comment (supakeen) [reply test/test_compile.py]: Ah yes, I added the `parser` argument I think because I had ideas about testing but they weren't very good. Let's remove that argument.

Perhaps there is a better way than my POSIX extension hacks to do the stdin and stdout handling. I did it this way so the code could just use `pathlib` everywhere. It'd be possible I guess to go with file-descriptors instead.

Comment (schuellerf) [reply test/test_compile.py]: If you don't mind I'll also change this within this PR as this would be needed to mock stdin I guess

Comment (mvo5) [reply test/test_compile.py]: Using sys.{stdin,stdout} seems fine with me, is there any downside? It's also easy to mock via pytest

File: test/test_compile.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  "test": 1,
+  "version": "2",
+  "sources": {}
+}""",
+            0,
+            None,
+        ),
+    ],
+)
+def test_compile(
```
Comment (mvo5) [inline test/test_compile.py:27]: Probably nice to remove the sys_exit_code and log_message as they are currently unused. and/or add a negative test that just makes sure validate fails and does not leave an (invalid/empty) output file behind.

File: test/test_compile.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    caplog,
+    capsys,
+    monkeypatch,
+    arguments,
+    input_data,
+    output_data,
+    sys_exit_code,
+    log_message,
+):
+    if arguments.input == "GENERATE":
```
Comment (mvo5) [inline test/test_compile.py:38]: (super nitpick) I wonder if the test parameter really needs to be a full argparse.Namespace, it seems we are only intrested in "file" or "stdin/stdout", maybe a combinartion. So just passing "input", "output" as either "some-filename" or "-" (for stdin/stdout) might be slightly shorter and then here we would construct the namespace

File: test/test_compile.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        input_filename = "input.yaml"
+        arguments.input = os.path.join(tmp_path, input_filename)
```
Comment (mvo5) [inline test/test_compile.py:40]: (super niptick) I like the way pathlib supports joins for paths ` arguments.input = tmp_path / input_filename` (but probably very personal and maybe does not work because input expects a string not a pathlib.Path?

**Review by mvo5 — COMMENTED** [review]
Thank you! I love how this looks nicer and nicer :) Some more ideas/suggestions (sorry, stop me if you feel it's too much)

**Review by mvo5 — APPROVED** [review]
Sorry, I meant to already approve in my previous comment .There are some ideas/suggestions left but it's IMHO a really nice step forwoard and we can do more in followups.

File: test/test_compile.py (hunk 4 above)
Comment (mvo5) [reply test/test_compile.py]: Let's keep as is for now and do the stdin/stdout testing in a followup. This way we can also rework it a bit, I think using filedescriptors instead of path has some benefits (including that it will make the sys.std{in,out} tests simpler and that tit will have less TOCTOU issues (which are not very important here)

Comment (schuellerf) [reply test/test_compile.py]: I don't think this would be TOCTOU...
please check my suggestion (next force push)

File: test/test_compile.py (hunk 9 above)
Comment (schuellerf) [reply test/test_compile.py:40]: oh - yea - I'm just no used to this pathlib feature and still feels strange to override operators like this but I'll change

**Review by supakeen — APPROVED** [review]
